File: src/dotnet/src/Dependabot.Job/Models/AllowedUpdate.cs
using System.Text.Json.Serialization;
using Dependabot.Job.Enums;

namespace Dependabot.Job;

/// <summary>
/// Represents an allowed update configuration for dependencies.
/// </summary>
public sealed record AllowedUpdate
{
  /// <summary>
  /// Gets or sets the type of update allowed.
  /// </summary>
  public UpdateType UpdateType { get; init; } = UpdateType.All;

  /// <summary>
  /// Gets or sets the dependency name pattern (supports wildcards).
  /// </summary>
  public string? DependencyName { get; init; }

  /// <summary>
  /// Gets or sets the dependency type.
  /// </summary>
  public string? DependencyType { get; init; }
}

File: src/ruby/lib/dependabot/job/models/allowed_update.rb
# typed: strict
# frozen_string_literal: true

require 'sorbet-runtime'
require_relative 'enums/update_type'

module Dependabot
  module Job
    module Models
      # Represents an allowed update configuration for dependencies.
      class AllowedUpdate < T::Struct
        extend T::Sig

        # The type of update allowed.
        const :update_type, Enums::UpdateType, default: Enums::UpdateType::All

        # The dependency name pattern (supports wildcards).
        const :dependency_name, T.nilable(String), default: nil

        # The dependency type.
        const :dependency_type, T.nilable(String), default: nil

        sig { params(hash: T::Hash[Symbol, T.untyped]).returns(AllowedUpdate) }
        def self.from_hash(hash)
          update_type = if hash[:update_type]
                          Enums::UpdateType.deserialize(hash[:update_type])
                        else
                          Enums::UpdateType::All
                        end

          new(
            update_type: update_type,
            dependency_name: hash[:dependency_name],
            dependency_type: hash[:dependency_type]
          )
        end

        sig { returns(T::Hash[Symbol, T.untyped]) }
        def to_hash
          {
            update_type: update_type.serialize,
            dependency_name: dependency_name,
            dependency_type: dependency_type
          }.compact
        end
      end
    end
  end
end
